components/tikv_util/src/topn.rs

```diff
@@ -0,0 +1,116 @@
+// Copyright 2021 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::cmp::Reverse;
+use std::collections::{binary_heap, BinaryHeap};
+use std::iter;
+
+// todo: Make it TopN<T, const N: usize> when we update our Rust version
+/// TopN is used to collect the largest `cap` items pushed in
+pub struct TopN<T> {
+    capacity: usize,
+    heap: BinaryHeap<Reverse<T>>,
+}
+
+impl<T: Ord> TopN<T> {
+    /// Create a `TopN` with fixed capacity.
+    pub fn new(capacity: usize) -> TopN<T> {
+        TopN {
+            capacity,
+            heap: BinaryHeap::with_capacity(capacity),
+        }
+    }
+
+    /// Pushes an `item` into `TopN`.
+    pub fn push(&mut self, item: T) {
+        self.heap.push(Reverse(item));
+        if self.heap.len() > self.capacity {
+            self.heap.pop();
+        }
+    }
+
+    /// Pops an `item` from `TopN`
+    pub fn pop(&mut self) -> Option<T> {
+        self.heap.pop().map(|Reverse(x)| x)
+    }
+
+    /// Returns the length of the `TopN`.
+    pub fn len(&self) -> usize {
+        self.heap.len()
+    }
+
+    /// Returns whether the `TopN` is empty.
+    pub fn is_empty(&self) -> bool {
+        self.heap.is_empty()
+    }
+
+    /// Returns the smallest item in the `TopN`, or `None` if it is empty.
+    pub fn peek(&self) -> Option<&T> {
+        self.heap.peek().map(|Reverse(x)| x)
+    }
+}
+
+impl<T> IntoIterator for TopN<T> {
+    type Item = T;
+
+    // this is added for rust-clippy#1013
+    #[allow(clippy::type_complexity)]
+    type IntoIter = iter::Map<binary_heap::IntoIter<Reverse<T>>, fn(Reverse<T>) -> T>;
+
+    // note: IntoIterator doesn't require the result in order, there is an `IntoIterSorted`, implement that if necessary
+    fn into_iter(self) -> Self::IntoIter {
+        self.heap.into_iter().map(|Reverse(x)| x)
+    }
+}
```

Comment:
How about just make `heap` public, so we can just `xxx_topn.heap.into_iter()`

Reply:
I tend not to expose `heap` in order to prevent `topn` from unexpected usage.